ad-aws-login.sh

```diff
@@ -79,31 +85,76 @@ function exit_error() {
 function cleanup() {
   (
     rm -f "${TEMP_FILE}" || true
-    kill "$(pgrep -lf "${USER_BROWSER}.app" | grep "${USER_BROWSER}.*--user-data-dir=${PWD}/user_data$" | awk '{print $1;}')"
+    kill "$BROWSER_PID" || true
   ) &>/dev/null
 }
 
-function handle_browser() {
+function read_os() {
+  case "$OSTYPE" in
+    solaris*) echo "SOLARIS";;
+    darwin*)  echo "OSX";; 
+    linux*)   echo "LINUX";;
+    bsd*)     echo "BSD";;
+    msys*)    echo "WIN";;
+    cygwin*)  echo "WIN";;
+  esac
+}
+
+function find_browser_path() {
+  local list_paths="$1"
+  local list_browsers="$2"
+  local path
   local browser
-  while read -r browser; do
-    if [[ -d "/Applications/${browser}.app" ]]; then
-      readonly USER_BROWSER="${browser}"
-      break
-    fi
-  done < <(echo "$BROWSERS")
+
+  while read -r path; do
+    while read -r browser; do
+      if [[ -f "${path}/${browser}" ]]; then
+        echo "${path}/${browser}"
+        break
+      fi
+    done < <(echo "$list_browsers")
+  done < <(echo "$list_paths")
   trap cleanup EXIT
+}
 
-  if [[ -z "${USER_BROWSER}" ]]; then
-    exit_error 1 "Cannot find a browser from:\n${BROWSERS}."
+function find_browser() {
+  local os=$(read_os)
+  case "$os" in
+    OSX*) echo $(find_browser_path "${FIND_PATHS_OSX}" "${BROWSERS_OSX}");;
+    WIN*) echo $(find_browser_path "${FIND_PATHS_WIN}" "${BROWSERS_WIN}");;
+    *)    exit_error 2 "Unhandled OS: ${OSTYPE}";;
+  esac
+}
+
+function open_browser() {
+  local os=$(read_os)
+  local browser=$(find_browser)
+  local url="https://myapps.microsoft.com"
+
+  if [[ -z "${browser}" ]]; then
+    exit_error 1 "Cannot find a browser"
   fi
 
-  args="--load-extension="${PWD}/chrome_extension" --disable-extensions-except="${PWD}/chrome_extension" --user-data-dir="${PWD}/user_data""
   # shellcheck disable=SC2086
-  open -a "${USER_BROWSER}" -F -n "https://myapps.microsoft.com" --args ${args}
+  args="--load-extension="${PWD}/chrome_extension" --disable-extensions-except="${PWD}/chrome_extension" --user-data-dir="${PWD}/user_data""
 
+  case "$os" in
+    OSX*)  open "${browser}" -F -n ${url} --args ${args};;
+    WIN*)  "${browser}" -F -n ${url} --args ${args} &;;
+    *)     exit_error 2 "Unhandled OS: ${OSTYPE}";;
+  esac
+
+  local pid=$!
+  echo $pid
+}
+
+function handle_browser() {
+  BROWSER_PID=$(open_browser)
+  
   while true; do
-    PID=$(pgrep -lf "${USER_BROWSER}\.app.*--user-data-dir=${PWD}/user_data$" | grep -v grep | awk '{print $1;}' || true)
-    if [[ -n "${PID}" ]]; then
+    local process=$(ps -p $BROWSER_PID | grep $BROWSER_PID)
```

Comment:
This will not work, since BROWSER_PID that was returned on line 147 contains the process id for open command, not the browser command. Browser will stay open for every. Therefore pgrep -lf "${USER_BROWSER}\.app.*--user-data-dir=${PWD}/user_data$" | grep -v grep | awk '{print $1;}' || true did the trick. Dirty but...

Reply:
Is there an option to launch the browser in the background the same way as in the POSIX terminals? e.g. `/abs/path/to/binary &` 
I don't mind restoring that pgrep somehow, but it's not available in the mingw shell in Windows by default.

Reply:
Separated the launching methods to Mac and Posix ones. Hopefully this way it works now for Mac as well.